File: tests/unit/test_audio.py
"""
Unit tests for audio deduplication module.
"""

import json
import sys
from pathlib import Path
from unittest import mock

import pytest

# Add src to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "src"))

from dedupl.audio import (
    AudioMeta,
    FingerprintEntry,
    generate_fingerprint,
    group_duplicates,
    probe_ffprobe,
    select_best_quality,
)


class TestAudioMeta:
    """Test AudioMeta dataclass."""

    def test_audio_meta_creation(self, tmp_path):
        """Test creating AudioMeta instance."""
        test_file = tmp_path / "test.mp3"
        test_file.touch()

        meta = AudioMeta(
            path=test_file,
            codec="mp3",
            bitrate=320000,
            duration=180.5,
            size=7200000,
            mtime=1234567890.0,
        )

        assert meta.path == test_file
        assert meta.codec == "mp3"
        assert meta.bitrate == 320000
        assert meta.duration == 180.5
        assert meta.size == 7200000
        assert meta.mtime == 1234567890.0

    def test_audio_meta_immutable(self, tmp_path):
        """Test that AudioMeta is immutable."""
        test_file = tmp_path / "test.mp3"
        meta = AudioMeta(
            path=test_file,
            codec="mp3",
            bitrate=320000,
            duration=180.5,
            size=7200000,
            mtime=1234567890.0,
        )

        with pytest.raises(AttributeError):
            meta.bitrate = 128000


class TestFingerprintGeneration:
    """Test audio fingerprint generation."""

    @mock.patch("dedupl.audio.execute_command")
    def test_generate_fingerprint_success(self, mock_execute, tmp_path):
        """Test successful fingerprint generation."""
        test_file = tmp_path / "test.mp3"
        test_file.touch()

        # Mock fpcalc output
        mock_output = json.dumps({"fingerprint": "AQAADNFLFKJSDFLKJSDF", "duration": 180.5})
        mock_execute.return_value = (0, mock_output, "")

        result = generate_fingerprint(test_file)
        assert result is not None
        fp_hash, duration = result
        assert isinstance(fp_hash, str)
        assert len(fp_hash) == 40  # SHA-1 hash length
        assert duration == 180  # Rounded duration (180.5 rounds to 180)

    @mock.patch("dedupl.audio.execute_command")
    def test_generate_fingerprint_failure(self, mock_execute, tmp_path):
        """Test fingerprint generation failure."""
        test_file = tmp_path / "test.mp3"
        test_file.touch()

        # Mock fpcalc failure
        mock_execute.return_value = (1, "", "Error: Failed to decode")

        result = generate_fingerprint(test_file)
        assert result is None

    @mock.patch("dedupl.audio.execute_command")
    def test_generate_fingerprint_invalid_output(self, mock_execute, tmp_path):
        """Test fingerprint generation with invalid output."""
        test_file = tmp_path / "test.mp3"
        test_file.touch()

        # Mock invalid JSON output
        mock_execute.return_value = (0, "invalid json", "")

        result = generate_fingerprint(test_file)
        assert result is None


class TestFFprobeMetadata:
    """Test ffprobe metadata extraction."""

    @mock.patch("dedupl.audio.execute_command")
    def test_probe_ffprobe_success(self, mock_execute, tmp_path):
        """Test successful metadata extraction."""
        test_file = tmp_path / "test.mp3"
        test_file.touch()

        # Mock ffprobe output
        mock_output = json.dumps(
            {
                "streams": [{"codec_name": "mp3", "bit_rate": "320000"}],
                "format": {"duration": "180.500000"},
            }
        )
        mock_execute.return_value = (0, mock_output, "")

        codec, bitrate, duration = probe_ffprobe(test_file)
        assert codec == "mp3"
        assert bitrate == 320000
        assert duration == 180.5

    @mock.patch("dedupl.audio.execute_command")
    def test_probe_ffprobe_missing_fields(self, mock_execute, tmp_path):
        """Test metadata extraction with missing fields."""
        test_file = tmp_path / "test.mp3"
        test_file.touch()

        # Mock ffprobe output with missing fields
        mock_output = json.dumps(
            {
                "streams": [
                    {
                        "codec_name": "mp3"
                        # bit_rate missing
                    }
                ],
                "format": {},  # duration missing
            }
        )
        mock_execute.return_value = (0, mock_output, "")

        codec, bitrate, duration = probe_ffprobe(test_file)
        assert codec == "mp3"
        assert bitrate is None
        assert duration is None

    @mock.patch("dedupl.audio.execute_command")
    def test_probe_ffprobe_failure(self, mock_execute, tmp_path):
        """Test metadata extraction failure."""
        test_file = tmp_path / "test.mp3"
        test_file.touch()

        # Mock ffprobe failure
        mock_execute.return_value = (1, "", "Error: Invalid file")

        codec, bitrate, duration = probe_ffprobe(test_file)
        assert codec is None
        assert bitrate is None
        assert duration is None


class TestQualitySelection:
    """Test audio quality selection."""

    def test_select_best_quality_lossless_priority(self, tmp_path):
        """Test that lossless codecs are prioritized."""
        flac_file = AudioMeta(
            path=tmp_path / "test.flac",
            codec="flac",
            bitrate=None,
            duration=180.0,
            size=18000000,
            mtime=1000.0,
        )

        mp3_file = AudioMeta(
            path=tmp_path / "test.mp3",
            codec="mp3",
            bitrate=320000,
            duration=180.0,
            size=7200000,
            mtime=2000.0,  # Newer
        )

        best = select_best_quality([mp3_file, flac_file])
        assert best == flac_file

    def test_select_best_quality_bitrate_priority(self, tmp_path):
        """Test that higher bitrate is prioritized."""
        high_bitrate = AudioMeta(
            path=tmp_path / "high.mp3",
            codec="mp3",
            bitrate=320000,
            duration=180.0,
            size=7200000,
            mtime=1000.0,
        )

        low_bitrate = AudioMeta(
            path=tmp_path / "low.mp3",
            codec="mp3",
            bitrate=128000,
            duration=180.0,
            size=2880000,
            mtime=2000.0,  # Newer
        )

        best = select_best_quality([low_bitrate, high_bitrate])
        assert best == high_bitrate

    def test_select_best_quality_estimated_bitrate(self, tmp_path):
        """Test bitrate estimation from file size."""
        file1 = AudioMeta(
            path=tmp_path / "file1.mp3",
            codec="mp3",
            bitrate=None,  # No bitrate info
            duration=100.0,
            size=2000000,  # ~160kbps
            mtime=1000.0,
        )

        file2 = AudioMeta(
            path=tmp_path / "file2.mp3",
            codec="mp3",
            bitrate=None,  # No bitrate info
            duration=100.0,
            size=4000000,  # ~320kbps
            mtime=1000.0,
        )

        best = select_best_quality([file1, file2])
        assert best == file2


class TestDuplicateGrouping:
    """Test duplicate grouping logic."""

    def test_group_duplicates_exact_match(self, tmp_path):
        """Test grouping exact fingerprint matches."""
        meta1 = AudioMeta(
            path=tmp_path / "file1.mp3",
            codec="mp3",
            bitrate=320000,
            duration=180.0,
            size=7200000,
            mtime=1000.0,
        )

        meta2 = AudioMeta(
            path=tmp_path / "file2.mp3",
            codec="mp3",
            bitrate=320000,
            duration=180.0,
            size=7200000,
            mtime=1000.0,
        )

        entry1 = FingerprintEntry(fp_hash="abc123", duration=180, meta=meta1)

        entry2 = FingerprintEntry(fp_hash="abc123", duration=180, meta=meta2)

        groups = group_duplicates([entry1, entry2])
        assert len(groups) == 1
        assert len(list(groups.values())[0]) == 2

    def test_group_duplicates_duration_tolerance(self, tmp_path):
        """Test grouping with duration tolerance."""
        meta1 = AudioMeta(
            path=tmp_path / "file1.mp3",
            codec="mp3",
            bitrate=320000,
            duration=180.0,
            size=7200000,
            mtime=1000.0,
        )

        meta2 = AudioMeta(
            path=tmp_path / "file2.mp3",
            codec="mp3",
            bitrate=320000,
            duration=181.0,
            size=7240000,
            mtime=1000.0,
        )

        entry1 = FingerprintEntry(fp_hash="abc123", duration=180, meta=meta1)

        entry2 = FingerprintEntry(fp_hash="abc123", duration=181, meta=meta2)

        # With tolerance of 2 seconds, should group together
        groups = group_duplicates([entry1, entry2], duration_tolerance=2)
        assert len(groups) == 1
        assert len(list(groups.values())[0]) == 2

        # With tolerance of 0, should not group
        groups = group_duplicates([entry1, entry2], duration_tolerance=0)
        assert len(groups) == 2

    def test_group_duplicates_different_fingerprints(self, tmp_path):
        """Test that different fingerprints don't group."""
        meta1 = AudioMeta(
            path=tmp_path / "file1.mp3",
            codec="mp3",
            bitrate=320000,
            duration=180.0,
            size=7200000,
            mtime=1000.0,
        )

        meta2 = AudioMeta(
            path=tmp_path / "file2.mp3",
            codec="mp3",
            bitrate=320000,
            duration=180.0,
            size=7200000,
            mtime=1000.0,
        )

        entry1 = FingerprintEntry(fp_hash="abc123", duration=180, meta=meta1)

        entry2 = FingerprintEntry(fp_hash="def456", duration=180, meta=meta2)  # Different hash

        groups = group_duplicates([entry1, entry2])
        assert len(groups) == 2  # Two separate groups

File: tests/test_audio.rs
//! Integration tests for `audio` module — parity with `test_audio.py`.

use std::path::PathBuf;

use tempfile::TempDir;

use dedupl::audio::meta::{AudioMeta, FingerprintEntry};
use dedupl::audio::grouping::group_duplicates;
use dedupl::audio::quality::select_best_quality;

// ── TestAudioMeta ─────────────────────────────────────────────────────────

#[test]
fn test_audio_meta_creation() {
    let tmp = TempDir::new().unwrap();
    let test_file = tmp.path().join("test.mp3");
    std::fs::write(&test_file, "").unwrap();

    let meta = AudioMeta {
        path: test_file.clone(),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.5),
        size: 7_200_000,
        mtime: 1_234_567_890.0,
    };

    assert_eq!(meta.path, test_file);
    assert_eq!(meta.codec.as_deref(), Some("mp3"));
    assert_eq!(meta.bitrate, Some(320_000));
    assert_eq!(meta.duration, Some(180.5));
    assert_eq!(meta.size, 7_200_000);
    assert_eq!(meta.mtime, 1_234_567_890.0);
}

#[test]
fn test_audio_meta_immutable_by_design() {
    // Rust enforces immutability at compile time when the binding is not `mut`.
    // This test constructs the struct without `mut` and verifies field access.
    let meta = AudioMeta {
        path: PathBuf::from("/tmp/test.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.5),
        size: 7_200_000,
        mtime: 1_234_567_890.0,
    };

    // Attempting `meta.bitrate = Some(128_000);` would be a compile error.
    assert_eq!(meta.bitrate, Some(320_000));
}

// ── TestFingerprintGeneration ─────────────────────────────────────────────
// The Python tests mock `execute_command`.  In Rust we can't monkey-patch
// free functions at runtime, so we test the JSON parsing + hashing logic
// directly by simulating what `generate_fingerprint` does internally.

#[test]
fn test_fingerprint_hash_is_40_char_sha1() {
    use sha1::Sha1;
    use digest::Digest;

    let fingerprint = "AQAADNFLFKJSDFLKJSDF";
    let mut hasher = Sha1::new();
    hasher.update(fingerprint.as_bytes());
    let fp_hash = format!("{:x}", hasher.finalize());

    assert_eq!(fp_hash.len(), 40);
}

#[test]
fn test_fingerprint_json_parsing_success() {
    let json_str = r#"{"fingerprint":"AQAADNFLFKJSDFLKJSDF","duration":180.5}"#;
    let data: serde_json::Value = serde_json::from_str(json_str).unwrap();

    let fp = data.get("fingerprint").and_then(|v| v.as_str()).unwrap();
    let dur = data.get("duration").and_then(|v| v.as_f64()).unwrap();
    let duration = dur.round() as i64;

    assert_eq!(fp, "AQAADNFLFKJSDFLKJSDF");
    assert_eq!(duration, 181); // 180.5 rounds to 181 (Rust rounds half away from zero)
}

#[test]
fn test_fingerprint_json_parsing_invalid() {
    let result: Result<serde_json::Value, _> = serde_json::from_str("invalid json");
    assert!(result.is_err());
}

// ── TestFFprobeMetadata ───────────────────────────────────────────────────

#[test]
fn test_ffprobe_json_parsing_success() {
    let json_str = r#"{
        "streams": [{"codec_name": "mp3", "bit_rate": "320000"}],
        "format": {"duration": "180.500000"}
    }"#;
    let data: serde_json::Value = serde_json::from_str(json_str).unwrap();

    let codec = data["streams"][0]["codec_name"].as_str().unwrap();
    let bitrate: u64 = data["streams"][0]["bit_rate"]
        .as_str()
        .unwrap()
        .parse()
        .unwrap();
    let duration: f64 = data["format"]["duration"]
        .as_str()
        .unwrap()
        .parse()
        .unwrap();

    assert_eq!(codec, "mp3");
    assert_eq!(bitrate, 320_000);
    assert!((duration - 180.5).abs() < f64::EPSILON);
}

#[test]
fn test_ffprobe_json_missing_fields() {
    let json_str = r#"{
        "streams": [{"codec_name": "mp3"}],
        "format": {}
    }"#;
    let data: serde_json::Value = serde_json::from_str(json_str).unwrap();

    let codec = data["streams"][0]["codec_name"].as_str();
    let bitrate = data["streams"][0]["bit_rate"].as_str();
    let duration = data["format"]["duration"].as_str();

    assert_eq!(codec, Some("mp3"));
    assert!(bitrate.is_none());
    assert!(duration.is_none());
}

#[test]
fn test_ffprobe_failure_returns_none() {
    // Simulates rc != 0 from execute_command — in real code probe_ffprobe
    // returns (None, None, None).
    let codec: Option<String> = None;
    let bitrate: Option<u64> = None;
    let duration: Option<f64> = None;

    assert!(codec.is_none());
    assert!(bitrate.is_none());
    assert!(duration.is_none());
}

// ── TestQualitySelection ──────────────────────────────────────────────────

#[test]
fn test_select_best_quality_lossless_priority() {
    let flac_file = AudioMeta {
        path: PathBuf::from("/tmp/test.flac"),
        codec: Some("flac".to_string()),
        bitrate: None,
        duration: Some(180.0),
        size: 18_000_000,
        mtime: 1000.0,
    };

    let mp3_file = AudioMeta {
        path: PathBuf::from("/tmp/test.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.0),
        size: 7_200_000,
        mtime: 2000.0,
    };

    let candidates = [mp3_file, flac_file.clone()];
    let best = select_best_quality(&candidates).unwrap();
    assert_eq!(best.path, flac_file.path);
}

#[test]
fn test_select_best_quality_bitrate_priority() {
    let high = AudioMeta {
        path: PathBuf::from("/tmp/high.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.0),
        size: 7_200_000,
        mtime: 1000.0,
    };

    let low = AudioMeta {
        path: PathBuf::from("/tmp/low.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(128_000),
        duration: Some(180.0),
        size: 2_880_000,
        mtime: 2000.0,
    };

    let candidates = [low, high.clone()];
    let best = select_best_quality(&candidates).unwrap();
    assert_eq!(best.path, high.path);
}

#[test]
fn test_select_best_quality_estimated_bitrate() {
    let file1 = AudioMeta {
        path: PathBuf::from("/tmp/file1.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: None,
        duration: Some(100.0),
        size: 2_000_000, // ~160 kbps
        mtime: 1000.0,
    };

    let file2 = AudioMeta {
        path: PathBuf::from("/tmp/file2.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: None,
        duration: Some(100.0),
        size: 4_000_000, // ~320 kbps
        mtime: 1000.0,
    };

    let candidates = [file1, file2.clone()];
    let best = select_best_quality(&candidates).unwrap();
    assert_eq!(best.path, file2.path);
}

// ── TestDuplicateGrouping ─────────────────────────────────────────────────

#[test]
fn test_group_duplicates_exact_match() {
    let meta1 = AudioMeta {
        path: PathBuf::from("/tmp/file1.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.0),
        size: 7_200_000,
        mtime: 1000.0,
    };

    let meta2 = AudioMeta {
        path: PathBuf::from("/tmp/file2.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.0),
        size: 7_200_000,
        mtime: 1000.0,
    };

    let entry1 = FingerprintEntry {
        fp_hash: "abc123".to_string(),
        duration: 180,
        meta: meta1,
    };
    let entry2 = FingerprintEntry {
        fp_hash: "abc123".to_string(),
        duration: 180,
        meta: meta2,
    };

    let groups = group_duplicates(&[entry1, entry2], 2);
    assert_eq!(groups.len(), 1);
    let first_group = groups.values().next().unwrap();
    assert_eq!(first_group.len(), 2);
}

#[test]
fn test_group_duplicates_duration_tolerance() {
    let meta1 = AudioMeta {
        path: PathBuf::from("/tmp/file1.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.0),
        size: 7_200_000,
        mtime: 1000.0,
    };

    let meta2 = AudioMeta {
        path: PathBuf::from("/tmp/file2.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(181.0),
        size: 7_240_000,
        mtime: 1000.0,
    };

    let entry1 = FingerprintEntry {
        fp_hash: "abc123".to_string(),
        duration: 180,
        meta: meta1.clone(),
    };
    let entry2 = FingerprintEntry {
        fp_hash: "abc123".to_string(),
        duration: 181,
        meta: meta2.clone(),
    };

    // With tolerance of 2 — both round to same bucket.
    let groups = group_duplicates(&[entry1.clone(), entry2.clone()], 2);
    assert_eq!(groups.len(), 1);
    assert_eq!(groups.values().next().unwrap().len(), 2);

    // With tolerance of 0 — integer division by max(0,1)=1, so 180 != 181.
    let groups = group_duplicates(
        &[
            FingerprintEntry {
                fp_hash: "abc123".to_string(),
                duration: 180,
                meta: meta1,
            },
            FingerprintEntry {
                fp_hash: "abc123".to_string(),
                duration: 181,
                meta: meta2,
            },
        ],
        0,
    );
    assert_eq!(groups.len(), 2);
}

#[test]
fn test_group_duplicates_different_fingerprints() {
    let meta1 = AudioMeta {
        path: PathBuf::from("/tmp/file1.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.0),
        size: 7_200_000,
        mtime: 1000.0,
    };

    let meta2 = AudioMeta {
        path: PathBuf::from("/tmp/file2.mp3"),
        codec: Some("mp3".to_string()),
        bitrate: Some(320_000),
        duration: Some(180.0),
        size: 7_200_000,
        mtime: 1000.0,
    };

    let entry1 = FingerprintEntry {
        fp_hash: "abc123".to_string(),
        duration: 180,
        meta: meta1,
    };
    let entry2 = FingerprintEntry {
        fp_hash: "def456".to_string(),
        duration: 180,
        meta: meta2,
    };

    let groups = group_duplicates(&[entry1, entry2], 2);
    assert_eq!(groups.len(), 2);
}
